File: crates/topcoat-runtime/src/runtime/interop/_f64.rs
use std::fmt::Write;
use topcoat_view::runtime::IntoViewParts;

use crate::runtime::Interop;

#[allow(non_camel_case_types)]
pub struct f64(core::primitive::f64);

impl Interop for core::primitive::f64 {
    type Surrogate = f64;

    fn to_js(&self, out: &mut String) {
        write!(out, "__context.builtin.f64({self})").unwrap();
    }

    fn into_surrogate(self) -> Self::Surrogate {
        f64(self)
    }
}

impl IntoViewParts for f64 {
    fn into_view_parts(self) -> impl Iterator<Item = topcoat_view::runtime::ViewPart> {
        self.0.into_view_parts()
    }
}

macro_rules! impl_binary_op {
    ($trait:ident, $method:ident, $op:tt) => {
        impl core::ops::$trait for f64 {
            type Output = f64;

            #[inline]
            fn $method(self, rhs: f64) -> f64 {
                f64(self.0 $op rhs.0)
            }
        }
    };
}

impl_binary_op!(Add, add, +);
impl_binary_op!(Sub, sub, -);
impl_binary_op!(Mul, mul, *);
impl_binary_op!(Div, div, /);

File: crates/topcoat-runtime/browser/src/surrogate/f64.ts
export class F64 {
	constructor(private readonly v: number) {}

	add(other: F64): F64 {
		return new F64(this.v + other.v);
	}

	sub(other: F64): F64 {
		return new F64(this.v - other.v);
	}

	mul(other: F64): F64 {
		return new F64(this.v * other.v);
	}

	div(other: F64): F64 {
		return new F64(this.v / other.v);
	}

	toString(): string {
		return this.v.toString();
	}
}
